### ISSUE TITLE
Drastic increase in build and warm up times when upgrading to 3.9.0

### ISSUE THREAD

**Issue by TimGuendel** [nuxt/nuxt#25072]
### Environment

```
- Operating System: Windows_NT
- Node Version:     v18.18.2
- Nuxt Version:     3.8.2
- CLI Version:      3.10.0
- Nitro Version:    2.8.1
- Package Manager:  yarn@3.6.4
- Builder:          -
- User Config:      telemetry, experimental, devtools, components, app, nitro, imports, vite, css, modules, pinia, image, proxy, runtimeConfig
- Runtime Modules:  @vueuse/nuxt@10.7.0, @pinia/nuxt@0.5.1, nuxt-jsonld@2.0.8, @nuxt-alt/proxy@2.4.9, @nuxt/image@1.1.0
- Build Modules:    -
```

### Reproduction

I tried, but it does not really show with an empty project 🤔
3.8.2 https://stackblitz.com/edit/github-r32mz1-hj5c5o?file=package.json
3.9.0 https://stackblitz.com/github/nuxt/starter/tree/v3/?file=package.json

### Describe the bug

My (large) project:
3.8.2: 
![image](https://github.com/nuxt/nuxt/assets/143795105/2e388987-b302-4318-bcd2-e0e2a1911dac)

3.9.0:
![image](https://github.com/nuxt/nuxt/assets/143795105/7a000b3c-732d-450f-a736-03666aa363b7)

Switching back to 3.8.2 leads to the shorter loading times again.

**Comment by manniL** [comment]
Would be nice to see if it is based on the amount of components or something else triggeres the high build time. Can you try to repro?

**Comment by danielroe** [comment]
We'd definitely want to tackle this if you can provide a reproduction.

Can you try installing vite v5 in your Nuxt 3.8.2 project, to see whether that's the cause? It should be completely compatible, but you'll need to use npm/pnpm/yarn resolutions/overrides.

**Comment by TimGuendel** [comment]
> We'd definitely want to tackle this if you can provide a reproduction.
> 
> Can you try installing vite v5 in your Nuxt 3.8.2 project, to see whether that's the cause? It should be completely compatible, but you'll need to use npm/pnpm/yarn resolutions/overrides.

I could not find any information on how to do this. Simply using ```"overrides": {
        "vite": "5.0.11"
    },``` did not do anything.

- How can I force Nuxt 3.8.2 to use Vite 5?
- How can I make sure it is indeed using Vite 5?

**Comment by TimGuendel** [comment]
All right, @manniL helped me 👍. Vite 5 seems to be the culprit:

![image](https://github.com/nuxt/nuxt/assets/143795105/8513f37c-f5f6-49e6-8b19-0f0d7934322b)

**Comment by danielroe** [comment]
It would be great to have a reproduction. Hopefully this is something we can solve in vite. 🙏

**Comment by danielroe** [comment]
One thing - when you set `vite.warmupEntry` to false, this should disappear (as it disables the 'in-advance' warmup). I'd be interested to know, however, if your initial browser request to your site is a similar length of time.

**Comment by TimGuendel** [comment]
> One thing - when you set `vite.warmupEntry` to false, this should disappear (as it disables the 'in-advance' warmup). I'd be interested to know, however, if your initial browser request to your site is a similar length of time.

So I did some tests and took screenshots of the nitro console and the browser timeline in the network tab. I waited for all chunks to be loaded:

3.8.2 with warmupEntry true
![image](https://github.com/nuxt/nuxt/assets/143795105/0cde9377-ffec-4cc6-be83-483b054309e7)
![image](https://github.com/nuxt/nuxt/assets/143795105/cdb4253a-7001-4829-bcd9-f72ef35664ad)

3.8.2 with warmupEntry false
![image](https://github.com/nuxt/nuxt/assets/143795105/c989e3b6-64c0-45c0-8dd7-0f5f12e0e9ae)
![image](https://github.com/nuxt/nuxt/assets/143795105/404ac1c3-4159-4791-8218-ea3e7655693a)

3.9.0 with warmupEntry true
![image](https://github.com/nuxt/nuxt/assets/143795105/f99b0126-5dab-43a0-83d6-06dc5da5e0c6)
![image](https://github.com/nuxt/nuxt/assets/143795105/e7b95197-f2af-4b6b-9746-122b558676c3)

3.9.0 with  warmupEntry false
![image](https://github.com/nuxt/nuxt/assets/143795105/f5c06484-31b4-4a86-8ad0-189362c025c9)
![image](https://github.com/nuxt/nuxt/assets/143795105/f4373090-8d1d-4af3-8fcb-32e9e8afaf16)

**Comment by danielroe** [comment]
cc: @antfu would love your thoughts on this

**Comment by antfu** [comment]
From the glance I can't really tell what's the cause. I think we still need a minimal reproduction to move forward

**Comment by rolf-koizumi-ecdb** [comment]
Not sure if this helps and unfortunately also no minimal reproduction, but I think I know why the warmup time is so much longer in our project with nuxt@3.9.1

First, I noticed that inside [warmupViteServer()](https://github.com/nuxt/nuxt/blob/6736decba0dc8bbc7276848a8a26973440ebdee4/packages/vite/src/utils/warmup.ts#L38C23-L38C39), nuxt@3.9.1 looks at more than double the amount of files than in nuxt@3.8.2 for our project.

I then looked at what is returned in [`deps`](https://github.com/nuxt/nuxt/blob/6736decba0dc8bbc7276848a8a26973440ebdee4/packages/vite/src/utils/warmup.ts#L66) in the warmup-function and noticed this difference:
```
// nuxt@3.8.2, with url = virtual:nuxt:/Users/[...]/.nuxt/routes.mjs
deps = [ '/pages/index.vue?macro=true' ]

// nuxt@3.9.1, with url = virtual:nuxt:/Users/[...]/.nuxt/routes.mjs 
deps = [
  ...,
  '/pages/imprint.vue?macro=true',
  '/pages/index.vue?macro=true',
  '/pages/login.vue?macro=true',
  ...
] // all pages of our project
```

And it seems that .nuxt/routes.mjs is warmed up a couple of times and every time, all pages are returned as dependencies.

Then, as an experiment I did this in [this line](https://github.com/nuxt/nuxt/blob/6736decba0dc8bbc7276848a8a26973440ebdee4/packages/vite/src/utils/warmup.ts#L67) in nuxt@3.9.1:
```js
if (url.includes('.nuxt/routes.mjs')) {
  await Promise.all(['/pages/index.vue?macro=true'].map((m) => warmup(m)));
} else {
  await Promise.all(deps.map((m) => warmup(m)));
}
```

With which the warmup time went down a lot (`Nitro built in 7977 ms` for me vs. `Nitro built in 1473 ms` in nuxt@3.8.2 without modifications).

I am still looking for a way to create a minimal reproduction out of this and I don't know if this is a problem or it is by design that nuxt@3.9.1 should look at all files multiple times at warmup. Maybe we should manually specify which files to warmup in the config, I will try that next.

**Comment by TimGuendel** [comment]
Was this helpful at all, or is there something else I could try to help solving this issue? 🤔

**Comment by TimGuendel** [comment]
@manniL suggested that I "empty" my templates and try again. We have +/-350 vue files and I replaced all templates by ```<template><div>Hello</div></template>```

These are the startup times:

<img width="360" alt="image" src="https://github.com/nuxt/nuxt/assets/143795105/f6b9f404-381b-407a-9327-401a032a6c78">

<img width="371" alt="image" src="https://github.com/nuxt/nuxt/assets/143795105/f6d1918c-5840-4d5e-b61d-414958c0f032">

Although significantly decreased, the build time with Nuxt 3.9.3 is almost double the time it took with Nuxt 3.8.2.
The warm up times also increased by a lot.

**Comment by TimGuendel** [comment]
So this issue seems to be resolved with 3.11.1... closing this.
